app/controllers/concerns/idv_session.rb

```diff
@@ -12,12 +12,37 @@ def confirm_idv_session_started
     redirect_to idv_session_url unless idv_session[:params].present?
   end
 
+  def confirm_idv_attempts_allowed
+    if idv_attempter.exceeded?
+      flash[:error] = t('idv.errors.hardfail')
+      redirect_to idv_fail_url
+    elsif idv_attempter.reset_attempts?
+      self.idv_attempts = 0
+    end
+  end
+
   def idv_session
     user_session[:idv] ||= {}
   end
 
   protected
 
+  def idv_attempts=(num)
+    current_user.update!(idv_attempts: num)
+  end
+
+  def idv_attempts
+    current_user.idv_attempts
+  end
+
+  def idv_attempter
+    @_attempter ||= IdvAttempter.new(current_user)
+  end
+
+  def idv_flag_user_attempt
+    current_user.update!(idv_attempted_at: Time.zone.now)
+  end
+
```

Comment:
I thought about what was bothering me about having these methods in the concern. It's that they represent a concept that is best encapsulated in its own class. Again, to separate concerns and reduce the responsibilities of the controller/concern.

I'm thinking this class would have instance methods such as `increment_idv_attempt`, `flag_idv_attempt_at_current_time`, and `reset_idv_attempts`. This would still allow for decoupling of the storage implementation while having this functionality defined in a specialized class, which would make it easier to unit test, and avoid doing things like `self.idv_attempts += 1` in the controller.

Thoughts?